### PR TITLE
Refactoring captcha code & captcha types as GLOBAL CONSTANTS

### LINKED ISSUE #4070: CAPTCHA code improvements
Not exactly related to this PR, but I think we should introduce some constants/static variables to define these captcha types rather than having these types of checks scattered throughout the codebase.

_Originally posted by @euantorano in https://github.com/mybb/mybb/pull/4003#discussion_r439869586_

<hr>

Not related to this PR as such, but this is one wild `if` branch. Would be a good idea to try and clean this up in a. future release.

_Originally posted by @euantorano in https://github.com/mybb/mybb/pull/4003#discussion_r439869231_

### PR DESCRIPTION
**Pull request by rajat315315**
Resolves #4070

### CODE REVIEW

File: admin/modules/config/settings.php [hunk 1]
```diff
@@ -958,18 +958,51 @@
...
+
+			if(in_array($captchaimage, array(NOCAPTCHA_RECAPTCHA, RECAPTCHA_INVISIBLE)) && (!$recaptchaprivatekey || !$recaptchapublickey)){
+				$set_captcha_image = true;
+			}
+			if(in_array($captchaimage, array(RECAPTCHA_V3)) && (!$recaptchaprivatekey || !$recaptchapublickey || !$recaptchascore)){
+				$set_captcha_image = true;
+			}
+			if(in_array($captchaimage, array(HCAPTCHA, HCAPTCHA_INVISIBLE)) && (!$hcaptchaprivatekey || !$hcaptchapublickey)){
+				$set_captcha_image = true;
+			}
```
Comment (JordanMussi) [inline admin/modules/config/settings.php]: Please could you make these `else if`s or as one `if` block.

File: admin/modules/config/settings.php [hunk 2]
```diff
@@ -958,18 +958,51 @@
...
-		   || (in_array($mybb->settings['captchaimage'], array(8)) && (!$mybb->settings['recaptchaprivatekey'] || !$mybb->settings['recaptchapublickey'] || !$mybb->settings['recaptchascore']))
-		   || (isset($mybb->input['upsetting']['captchaimage']) && in_array($mybb->input['upsetting']['captchaimage'], array(6, 7)) && (!$mybb->input['upsetting']['hcaptchaprivatekey'] || !$mybb->input['upsetting']['hcaptchapublickey']))
-		   || (in_array($mybb->settings['captchaimage'], array(6, 7)) && (!$mybb->settings['hcaptchaprivatekey'] || !$mybb->settings['hcaptchapublickey'])))
-		{
-			$mybb->input['upsetting']['captchaimage'] = 1;
+		require_once MYBB_ROOT.'inc/class_captcha.php';
+		
+		// Have we opted for a reCAPTCHA or hCaptcha and not set a public/private key in input?
+		$set_captcha_image = false;
+		if(isset($mybb->input['upsetting']['captchaimage'])){
```
Comment (JordanMussi) [inline admin/modules/config/settings.php]: Sorry for the bureaucracy but the 1.8 development standards require the opening brace on the next line, here and elsewhere in the PR.

https://docs.mybb.com/1.8/development/standards/#control-structures

File: admin/modules/config/settings.php [hunk 3]
```diff
@@ -958,18 +958,51 @@
...
+
+		if(in_array($captchaimage, array(NOCAPTCHA_RECAPTCHA, RECAPTCHA_INVISIBLE)) && (!$recaptchaprivatekey || !$recaptchapublickey)){
+			$set_captcha_image = true;
+		}
+		if(in_array($captchaimage, array(RECAPTCHA_V3)) && (!$recaptchaprivatekey || !$recaptchapublickey || !$recaptchascore)){
+			$set_captcha_image = true;
+		}
+		if(in_array($captchaimage, array(6, HCAPTCHA_INVISIBLE)) && (!$hcaptchaprivatekey || !$hcaptchapublickey)){
+			$set_captcha_image = true;
+		}
```
Comment (JordanMussi) [inline admin/modules/config/settings.php]: `else if`s or one `if` as above.

File: admin/modules/config/settings.php [hunk 4]
```diff
@@ -958,18 +958,51 @@
...
+		if($set_captcha_image){
+			$mybb->input['upsetting']['captchaimage'] = DEFAULT_CAPTCHA;
 			$lang->success_settings_updated .= $lang->success_settings_updated_captchaimage;
 		}
 
+
 		// If using fulltext then enforce minimum word length given by database
```
Comment (JordanMussi) [inline admin/modules/config/settings.php]:
```suggestion
		}

		// If using fulltext then enforce minimum word length given by database
```

Please undo this extra line insert.

File: inc/class_captcha.php [hunk 5]
```diff
@@ -33,6 +33,13 @@
  *
  */
 
+define('DEFAULT_CAPTCHA', 1);
```
Comment (JordanMussi) [inline inc/class_captcha.php]: How about `MYBB_DEFAULT` (when combined with the suggestion to make them contained within the class)?

File: inc/class_captcha.php [hunk 6]
```diff
@@ -33,6 +33,13 @@
...
+define('NOCAPTCHA_RECAPTCHA', 4);
+define('RECAPTCHA_INVISIBLE', 5);
+define('HCAPTCHA', 6);
+define('HCAPTCHA_INVISIBLE', 7);
+define('RECAPTCHA_V3', 8);
+
```
Comment (JordanMussi) [inline inc/class_captcha.php]: Personally I think these would be best placed as constants within the `captcha` class to emphasise the need to `require` this file.

Thoughts @euantorano?

File: inc/class_captcha.php [hunk 7]
```diff
@@ -107,30 +114,30 @@ function __construct($build = false, $template = "")
...
-			elseif($this->type == 8)
+			elseif($this->type == HCAPTCHA_INVISIBLE)
 			{
 				$this->captcha_template .= "_recaptcha_invisible";
 			}
 		}
 
 		// Work on which CAPTCHA we've got installed
-		if(in_array($this->type, array(4, 5, 8)) && $mybb->settings['recaptchapublickey'] && $mybb->settings['recaptchaprivatekey'])
+		if(in_array($this->type, array(NOCAPTCHA_RECAPTCHA, RECAPTCHA_INVISIBLE, HCAPTCHA_INVISIBLE)) && $mybb->settings['recaptchapublickey'] && $mybb->settings['recaptchaprivatekey'])
```
Comment (JordanMussi) [inline inc/class_captcha.php]:
```suggestion
		if(in_array($this->type, array(NOCAPTCHA_RECAPTCHA, RECAPTCHA_INVISIBLE, RECAPTCHA_V3)) && $mybb->settings['recaptchapublickey'] && $mybb->settings['recaptchaprivatekey'])
```

**Review by JordanMussi — CHANGES_REQUESTED** [review]
Thanks for tackling this! Left a few comments 😃 

I'm undecided on whether it's best to have one if statement with the conditions broken over several lines to avoid the temporary variables in `admin/modules/config/settings.php` but I think I'm edging towards one statement.

File: inc/class_captcha.php (hunk 6 above)
Comment (euantorano) [reply inc/class_captcha.php]: Agreed, that also fits more with the code standards we are going for with 1.9 and ahead where one file does one thing (e.g. define a class).

File: inc/class_captcha.php (hunk 7 above)
Comment (rajat315315) [reply inc/class_captcha.php]: ya, thanks, that surely is a mistake.

**Comment by euantorano** [comment]
This looks like a massive improvement to me - makes the code **much** easier to follow, thanks @rajat315315! I'm happy to merge this, but just want to make sure @JordanMussi is happy first?

**Comment by euantorano** [comment]
I'm going to go ahead and merge having heard no other feedback as of yet.

**Comment by dvz** [comment]
There are leftover constant references without the class prefix in `class_captcha.php`, `newreply.php`, `member.php` and `newthread.php`.

**Comment by rajat315315** [comment]
Hello everyone,
Its a sincere request to nominate me, if you think I have done some meaningful contribution to the community...
https://stars.github.com/nominate/

**Pull request merged**